srtcore/queue.cpp

```diff
@@ -216,24 +216,28 @@ CUnit *CUnitQueue::getNextAvailUnit()
     if (m_iCount >= m_iSize)
         return NULL;
 
-    CQEntry *entrance = m_pCurrQueue;
-
+    int units_checked = 0;
     do
     {
-        for (CUnit *sentinel = m_pCurrQueue->m_pUnit + m_pCurrQueue->m_iSize - 1; m_pAvailUnit != sentinel;
-             ++m_pAvailUnit)
+        CUnit *lastunit = m_pCurrQueue->m_pUnit + m_pCurrQueue->m_iSize - 1;
+        for (; m_pAvailUnit != lastunit; ++m_pAvailUnit, ++units_checked)
+        {
             if (m_pAvailUnit->m_iFlag == CUnit::FREE)
+            {
                 return m_pAvailUnit;
+            }
+        }
 
-        if (m_pCurrQueue->m_pUnit->m_iFlag == CUnit::FREE)
+        if (lastunit->m_iFlag == CUnit::FREE)
         {
-            m_pAvailUnit = m_pCurrQueue->m_pUnit;
+            m_pAvailUnit = lastunit;
```

Comment:
This condition should actually check the last unit in the queue, not the first one.

Reply:
But this doesn't make any sense either. If you make `lastunit` the `end` (past-the-end), this check would be included in the loop. In this loop the very last unit is skipped, and then you check it again. Of course, checking the first one doesn't make any sense at all because it looks like a dead code, if I get it right - this unit will be checked in the first step in the above loop, and will either result with returning it before this check can be done, or will pass on without finding anything and making this check again on the 1st unit, which will result to be always false.

Reply:
This is one point I wanted your thoughts on.

The best way would be to have just one loop:

```c++
CUnit *end = m_pCurrQueue->m_pUnit + m_pCurrQueue->m_iSize;
for (; m_pAvailUnit != end; ++m_pAvailUnit, ++units_checked)
{
   // ...
```

But this would make static analyzer to complain that `end` points out of bounds of the array.

Reply:
Kinda strange because past-the-end is a normal practice that even has even a special hookup in the C++ standard. I understand when it complains when you are trying to dereference the pointer.

Reply:
Let's go with `end` proposed above and bother with static check later then.